File: src/codex/PresentationProtocol.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later OR MIT

#include "codex/PresentationProtocol.h"

#include <utility>

namespace codexui::codex::presentation {
namespace {

nlohmann::json baseFrame(std::string kind) {
  return {{"protocol", ProtocolName},
          {"version", ProtocolVersion},
          {"kind", std::move(kind)}};
}

void addAuthorityAndScope(nlohmann::json &frame, Authority authority,
                          nlohmann::json scope) {
  frame["authority"] = authorityName(authority);
  if (scope.is_object() && !scope.empty())
    frame["scope"] = std::move(scope);
}

} // namespace

std::string_view authorityName(Authority authority) noexcept {
  switch (authority) {
  case Authority::Merge:
    return "merge";
  case Authority::Replace:
    return "replace";
  case Authority::Remove:
    return "remove";
  case Authority::None:
    return "none";
  }
  return "none";
}

nlohmann::json command(std::string action, nlohmann::json data,
                       std::string correlationId) {
  nlohmann::json frame = baseFrame("command");
  frame["action"] = std::move(action);
  frame["data"] = std::move(data);
  if (!correlationId.empty())
    frame["correlationId"] = std::move(correlationId);
  return frame;
}

nlohmann::json result(std::uint64_t sequence, std::uint64_t generation,
                      std::string action, std::string correlationId, bool ok,
                      nlohmann::json data, Authority authority,
                      nlohmann::json scope) {
  nlohmann::json frame = baseFrame("result");
  frame["sequence"] = sequence;
  frame["generation"] = generation;
  frame["action"] = std::move(action);
  frame["correlationId"] = std::move(correlationId);
  frame["ok"] = ok;
  frame[ok ? "data" : "error"] = std::move(data);
  addAuthorityAndScope(frame, authority, std::move(scope));
  return frame;
}

nlohmann::json event(std::uint64_t sequence, std::uint64_t generation,
                     std::string type, nlohmann::json data, Authority authority,
                     nlohmann::json scope) {
  nlohmann::json frame = baseFrame("event");
  frame["sequence"] = sequence;
  frame["generation"] = generation;
  frame["type"] = std::move(type);
  frame["data"] = std::move(data);
  addAuthorityAndScope(frame, authority, std::move(scope));
  return frame;
}

bool isPresentationFrame(const nlohmann::json &value) noexcept {
  if (!value.is_object())
    return false;
  const auto protocol = value.find("protocol");
  if (protocol == value.end() || !protocol->is_string() ||
      protocol->get_ref<const nlohmann::json::string_t &>() != ProtocolName)
    return false;
  const auto version = value.find("version");
  if (version == value.end() || !version->is_number_unsigned() ||
      version->get_ref<const nlohmann::json::number_unsigned_t &>() !=
          ProtocolVersion)
    return false;
  const auto kind = value.find("kind");
  if (kind == value.end() || !kind->is_string())
    return false;
  const std::string &kindValue =
      kind->get_ref<const nlohmann::json::string_t &>();
  const auto stringField = [&value](const char *name) {
    const auto member = value.find(name);
    return member != value.end() && member->is_string() &&
           !member->get_ref<const nlohmann::json::string_t &>().empty();
  };
  if (kindValue == "command") {
    const auto data = value.find("data");
    return stringField("action") && data != value.end() && data->is_object();
  }
  if (kindValue != "event" && kindValue != "result")
    return false;
  const auto sequence = value.find("sequence");
  const auto generation = value.find("generation");
  const auto authority = value.find("authority");
  if (sequence == value.end() || !sequence->is_number_unsigned() ||
      generation == value.end() || !generation->is_number_unsigned() ||
      authority == value.end() || !authority->is_string())
    return false;
  const std::string &authorityValue =
      authority->get_ref<const nlohmann::json::string_t &>();
  if (authorityValue != "none" && authorityValue != "merge" &&
      authorityValue != "replace" && authorityValue != "remove")
    return false;
  const auto scope = value.find("scope");
  if (scope != value.end() && !scope->is_object())
    return false;
  if (kindValue == "event") {
    const auto data = value.find("data");
    return stringField("type") && data != value.end() && data->is_object();
  }
  const auto ok = value.find("ok");
  if (!stringField("action") || !stringField("correlationId") ||
      ok == value.end() || !ok->is_boolean())
    return false;
  return ok->get_ref<const nlohmann::json::boolean_t &>()
             ? value.contains("data") && !value.contains("error")
             : value.contains("error") && !value.contains("data");
}

std::string stringMember(const nlohmann::json &value, const char *name) {
  if (!value.is_object())
    return {};
  const auto iterator = value.find(name);
  return iterator != value.end() && iterator->is_string()
             ? iterator->get<std::string>()
             : std::string{};
}

nlohmann::json member(const nlohmann::json &value, const char *name,
                      nlohmann::json fallback) {
  if (!value.is_object())
    return fallback;
  const auto iterator = value.find(name);
  return iterator == value.end() ? std::move(fallback) : *iterator;
}

} // namespace codexui::codex::presentation

File: web/src/presentation/PresentationProtocol.ts
export const PresentationProtocolName = "codexui.presentation";
export const PresentationProtocolVersion = 1;

export type Authority = "none" | "merge" | "replace" | "remove";
export type JsonObject = Record<string, unknown>;
export const ThreadSettingFields = ["activePermissionProfile", "approvalPolicy", "approvalsReviewer",
    "collaborationMode", "cwd", "effort", "instructionSources", "model", "modelProvider", "personality",
    "reasoningEffort", "sandbox", "sandboxPolicy", "serviceTier", "summary"] as const;
const ThreadSettingConceptAliases: Record<string, string> = {activePermissionProfile: "permissionProfile",
    approvalPolicy: "approval", approvalsReviewer: "reviewer", collaborationMode: "collaboration",
    permissions: "permissionProfile", reasoningEffort: "effort", sandboxPolicy: "sandbox"};

export interface ThreadSettingStamp {generation: number; sequence: number; acknowledgements: number}

export function threadSettingConcept(field: string): string {
    const alias = ThreadSettingConceptAliases[field];
    return alias ?? ((ThreadSettingFields as readonly string[]).includes(field) ? field : "");
}

export interface PresentationFrame extends JsonObject {
    protocol: typeof PresentationProtocolName;
    version: typeof PresentationProtocolVersion;
    kind: "command" | "event" | "result";
}

function baseFrame(kind: PresentationFrame["kind"]): PresentationFrame {
    return {
        protocol: PresentationProtocolName,
        version: PresentationProtocolVersion,
        kind,
    };
}

function addAuthorityAndScope(
    frame: PresentationFrame,
    authority: Authority,
    scope: JsonObject,
): void {
    frame.authority = authority;
    if (Object.keys(scope).length > 0) frame.scope = scope;
}

export function command(
    action: string,
    data: JsonObject = {},
    correlationId = "",
): PresentationFrame {
    const frame = baseFrame("command");
    frame.action = action;
    frame.data = data;
    if (correlationId !== "") frame.correlationId = correlationId;
    return frame;
}

export function result(
    sequence: number,
    generation: number,
    action: string,
    correlationId: string,
    ok: boolean,
    data: unknown,
    authority: Authority = "none",
    scope: JsonObject = {},
): PresentationFrame {
    const frame = baseFrame("result");
    frame.sequence = sequence;
    frame.generation = generation;
    frame.action = action;
    frame.correlationId = correlationId;
    frame.ok = ok;
    frame[ok ? "data" : "error"] = data;
    addAuthorityAndScope(frame, authority, scope);
    return frame;
}

export function event(
    sequence: number,
    generation: number,
    type: string,
    data: JsonObject = {},
    authority: Authority = "none",
    scope: JsonObject = {},
): PresentationFrame {
    const frame: PresentationFrame = {protocol: PresentationProtocolName, version: PresentationProtocolVersion,
        kind: "event", sequence, generation, type, data};
    addAuthorityAndScope(frame, authority, scope);
    return frame;
}

export function isObject(value: unknown): value is JsonObject {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function jsonEqual(left: unknown, right: unknown): boolean {
    if (Object.is(left, right)) return true;
    if (Array.isArray(left) || Array.isArray(right)) return Array.isArray(left) && Array.isArray(right)
        && left.length === right.length && left.every((value, index) => jsonEqual(value, right[index]));
    if (!isObject(left) || !isObject(right)) return false;
    const keys = Object.keys(left);
    return keys.length === Object.keys(right).length
        && keys.every(key => Object.hasOwn(right, key) && jsonEqual(left[key], right[key]));
}

function isUnsigned(value: unknown): value is number {
    return typeof value === "number" && Number.isInteger(value) && value >= 0;
}

export function isPresentationFrame(value: unknown): value is PresentationFrame {
    if (!isObject(value)) return false;
    if (
        value.protocol !== PresentationProtocolName
        || value.version !== PresentationProtocolVersion
        || typeof value.kind !== "string"
    ) {
        return false;
    }
    if (value.kind === "command") {
        return stringMember(value, "action") !== "" && isObject(value.data);
    }
    if (value.kind !== "event" && value.kind !== "result") return false;
    if (
        !isUnsigned(value.sequence)
        || !isUnsigned(value.generation)
        || (value.authority !== "none"
            && value.authority !== "merge"
            && value.authority !== "replace"
            && value.authority !== "remove")
        || (Object.hasOwn(value, "scope") && !isObject(value.scope))
    ) {
        return false;
    }
    if (value.kind === "event") {
        return stringMember(value, "type") !== "" && isObject(value.data);
    }
    if (
        stringMember(value, "action") === ""
        || stringMember(value, "correlationId") === ""
        || typeof value.ok !== "boolean"
    ) {
        return false;
    }
    return value.ok
        ? Object.hasOwn(value, "data") && !Object.hasOwn(value, "error")
        : Object.hasOwn(value, "error") && !Object.hasOwn(value, "data");
}

export function stringMember(value: unknown, name: string): string {
    return isObject(value) && typeof value[name] === "string"
        ? value[name]
        : "";
}

export function member(
    value: unknown,
    name: string,
    fallback: unknown = null,
): unknown {
    return isObject(value) && Object.hasOwn(value, name) ? value[name] : fallback;
}
